pkg/proxy/metaproxier/meta_proxier.go

```diff
@@ -98,6 +112,23 @@ func (proxier *metaProxier) OnServiceSynced() {
 	proxier.ipv6Proxier.OnServiceSynced()
 }
 
+// serviceIPFamily that returns IPFamily of service or error if
+// failed to identify the IP family.
+func serviceIPFamily(service *v1.Service) (*v1.IPFamily, error) {
```

Comment:
I guess this needs to look at `IPFamily` as well (to deal with headless services), and needs to fall back rather than erroring out if neither `ClusterIP` nor `IPFamily` is usable. Basically, this should be the same as `endpointutil.IsIPv6Service` in https://github.com/kubernetes/kubernetes/pull/91399/commits/ad313391 and maybe we should put that somewhere they can share it?

Reply:
that brings me one question, does kube-proxy need to care about headless or externalName service types?
I mean, does kube-proxy does something for a service that does not have the clusterIP set?

Reply:
oh, duh, I was thinking "headless services still have endpoints so it needs to care", but that's wrong. I think you're right. You'll need to look through the code to confirm that

Reply:
heh, interesting exercise. check the stack trace :)

OnServiceAdd() -> OnServiceUpdate() -> proxier.serviceChanges.Update()

https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/iptables/proxier.go#L536-L548

proxier.serviceChanges.Update() -> serviceToServiceMap()

https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/service.go#L241-L262

serviceToServiceMap() -> utilproxy.ShouldSkipService()
	https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/service.go#L306-L313
	
utilproxy.ShouldSkipService() -> IsServiceIPSet()
	https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/proxy/util/utils.go#L148-L161

and  IsServiceIPSet()	
	https://github.com/kubernetes/kubernetes/blob/6a5f0e6eda5d4d736adaa2b3dd08c4366153bacf/pkg/apis/core/helper/helpers.go#L267-L271